app/controllers/authenticate_controller.rb

```diff
@@ -1,90 +1,30 @@
 class AuthenticateController < ApplicationController
-  include TokenGenerator
-
-  class MemoizedRole
-    def self.[](role_id)
-      @user_roles ||= Hash.new { |h, id| h[id] = Role[id] }
-      @user_roles[role_id]
-    end
-
-    def self.roleid_from_username(account, username)
-      Role.roleid_from_username(account, username)
-    end
-  end
 
   def authenticate
 
-    authenticator  = params[:authenticator]
-    service_id     = params[:service_id]
-    account        = params[:account]
-    username       = params[:id]
-    password       = request.body.read
-    role_id        = MemoizedRole.roleid_from_username(account, username)
-
-    validate_roleid_is_nonempty!(role_id)
-    
-    case authenticator
-    when 'authn-ldap'
-      validate_security!(authenticator, account, service_id, username)
-      raise Unauthorized unless ldap_authenticator.valid?(username, password)
-      role = MemoizedRole[role_id]
-    when 'authn' # default conjur auth
-      credentials = Credentials[role_id]
-      validate_credentials!(credentials, password)
-      role = credentials.role
-    else
-      raise Unauthorized
-    end
-
-    # If we arrive here, the user is authenticated and `role` is set
-    # Otherwise Unauthorized would have already been raised
-
-    authentication_token = sign_token(role)
+    # TODO move the Strategy into an initializer
+    #
+    authentication_token = ::Authentication::Strategy.new(
+      # TODO: remove these (defaults work)
+      authenticators: ::Authentication::InstalledAuthenticators.new(ENV),
+      security: nil,
+      env: ENV,
+      role_class: ::Authentication::MemoizedRole,
+      token_factory: TokenFactory.new
+    ).conjur_token(
+      ::Authentication::Strategy::Input.new(
+        authenticator_name: params[:authenticator],
+        service_id:         params[:service_id],
+        account:            params[:account],
+        username:           params[:id],
+        password:           request.body.read
+      )
+    )
     render json: authentication_token
-  end
-  
-  protected
-
-  def validate_security!(authenticator, account, service_id, user_id)
-    security = Authenticators::Security.new(authn_type: authenticator,
-      account: account, role_class: MemoizedRole)
-    security.validate(service_id, user_id)
-  rescue AuthenticatorNotEnabled, ServiceNotDefined, NotAuthorizedInConjur => e
-    logger.debug(e.message)
-    raise Unauthorized
   rescue => e
-    logger.debug("Unexpected Authenticators::Security Error: #{e.message}")
+    puts '*******************************'
```

Comment:
What do you think about using the application `logger` for output?

Reply:
Yeah probably best.  I'll make that change.  I think eventually we should return json error response bodies, but that's a different update.